File: python/Swc/ui/subBackground.py
__author__ = 'Su Lei'

import cv2 as cv
import numpy as np
import math
import opsOfSub as op
from copy import deepcopy
import time


def subBackground(inImage):
    raw_image = inImage
    raw_image = np.float32(raw_image)

    height, width = raw_image.shape
    radius = 50
    ball, shrinkFactor = op.createRollingBall(radius)
    shrink = shrinkFactor > 1

    op.filter3x3(raw_image, 'mean')
    smoothed_image = deepcopy(raw_image)
    minmax = op.getMinMax(smoothed_image)

    if shrink:
        smallImage = op.shrinkImage(smoothed_image, shrinkFactor)
    else:
        smallImage = smoothed_image

    op.rollBall(ball, smallImage)
    outImage = op.enlargeImage(smallImage, height, width, shrinkFactor)
    outImage = np.uint8(outImage)
    raw_image = np.uint8(raw_image)
    outImage = raw_image - outImage
    return outImage





if __name__ == '__main__':
     raw_image = cv.imread('G:\\sulei\\workplace\\Swc\\test\\1.tif', cv.CV_LOAD_IMAGE_UNCHANGED)
     raw_image = np.float32(raw_image)
     t1 = time.time()
     height, width = raw_image.shape
     radius = 50
     ball, shrinkFactor = op.createRollingBall(radius)
     shrink = shrinkFactor > 1
     op.filter3x3(raw_image, 'mean')
     smoothed_image = deepcopy(raw_image)
     minmax = op.getMinMax(smoothed_image)
     if shrink:
         smallImage = op.shrinkImage(smoothed_image, shrinkFactor)
     else:
         smallImage = smoothed_image
         
     op.rollBall(ball, smallImage)
#     cv.imwrite('2.tif', smallImage)
     outImage = op.enlargeImage(smallImage, height, width, shrinkFactor)
     outImage = np.uint8(outImage)
#     cv.imwrite('3.tif', outImage)
     raw_image = np.uint8(raw_image)
     outImage = raw_image - outImage
     t2 = time.time()
     print 'main: ', t2 - t1
     cv.imwrite('2.tif', outImage)

File: matlab/substractBackground/subBackground.m
% function [outImage] = subBackground(inImage)

clc;
clear all;

% raw_image = inImage;
% raw_image = imread('G:\sulei\workplace\vs\subbackground\subbackground\1.tif');
% raw_image = double(raw_image);
% 
% 
% [height, width] = size(raw_image);
% radius = 50;
% [ball, shrinkFactor] = createRollingBall(radius);
% shrink = shrinkFactor > 1;
% 
% doPresmooth = true;
% if doPresmooth
%     smooted_image = filter3x3(raw_image, 'mean');
% end
% 
% 
% 
% minmax = getMinMax(smooted_image);
% 
% if shrink
%     smallImage = shrinkImage(smooted_image, shrinkFactor);
% else
%     smallImage = smooted_image;
% end
% 
% smallImage = rollBall(ball, smallImage);
% a = uint8(smallImage);
% imwrite(a,'1_sm.tif');
% outImage = enlargeImage(smallImage, [height, width], shrinkFactor);
% outImage = uint8(outImage);
% raw_image = uint8(raw_image);
% outImage = raw_image - outImage;
% imwrite(outImage,'1_bg1.tif');
    